ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala

```diff
@@ -0,0 +1,137 @@
+    executionContext: ExecutionContext
+) extends PagedCompletionsReader {
+
+  protected val cacheRef: AtomicReference[RangeCache[CompletionStreamResponseWithParties]] =
+    new AtomicReference(RangeCache.empty[CompletionStreamResponseWithParties](maxItems))
+
+  private val pendingRequestRef: AtomicReference[Future[Unit]] =
+    new AtomicReference[Future[Unit]](Future.unit)
+
+  /** Returns completions stream filtered based on parameters.
+    * Newest completions read from database are added to cache
+    *
+    * @param requestedRange requested offsets range
+    * @param applicationId completions application id
+    * @param parties required completions submitters
+    * @param loggingContext logging context
+    * @return
+    */
+  override def getCompletionsPage(
+      requestedRange: Range,
+      applicationId: ApplicationId,
+      parties: Set[Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] = {
+    val cache = cacheRef.get()
+    val historicCompletionsFuture =
+      requestedRange
+        .lesserRangeDifference(cache.range) // calculate historic range to fetch
+        .map(historicRange => fetchHistoric(historicRange, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val futureCompletionsFuture = requestedRange
+      .greaterRangeDifference(cache.range) // calculate future range to fetch
+      .map(futureRange => fetchFuture(futureRange, applicationId, parties))
+      .getOrElse(futureEmptyList)
+
+    val cachedCompletions = cache
+      .slice(requestedRange)
+      .map(slicedCache => filterAndMapToResponse(slicedCache.cache.toSeq, applicationId, parties))
+      .getOrElse(Seq.empty)
+    for {
+      historicCompletions <- historicCompletionsFuture
+      futureCompletions <- futureCompletionsFuture
+    } yield historicCompletions ++ cachedCompletions ++ futureCompletions
+  }
+
+  /** fetches completions older than start offset of cache
+    */
+  private def fetchHistoric(
+      range: Range,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] =
+    completionsDao.getFilteredCompletions(range, applicationId, parties)
+
+  /** fetches completions ahead cache and caches results
+    */
+  private def fetchFuture(
+      rangeToFetch: Range,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] =
+    pendingRequestRef.synchronized {
+      val pendingRequest = pendingRequestRef.get()
+      if (pendingRequest.isCompleted) {
+        val allCompletionsFuture = createAndSetNewPendingRequest(rangeToFetch)
+        allCompletionsFuture.map(filterAndMapToResponse(_, applicationId, parties))
+      } else {
+        pendingRequest.flatMap { _ =>
```

Comment:
I could not find a unit test for a situation where two or three `fetchFuture` are executed in a short succession. I.e. subsequent requests hit while the original one is being processed. Please note that such an occurrence is very likely to happen in a real world scenario, where Dao has a non-negligible latency.